rules-reviewed/technology-usage/database-technology-usage.windup.xml

```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0"?>
+<ruleset id="technology-usage-database" xmlns="http://windup.jboss.org/schema/jboss-ruleset" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://windup.jboss.org/schema/jboss-ruleset http://windup.jboss.org/schema/jboss-ruleset/windup-jboss-ruleset.xsd">
+    <metadata>
+        <description>
+            This ruleset provides statistical summaries of the database items that were found during the analysis.
+        </description>
+        <dependencies>
+            <addon id="org.jboss.windup.rules,windup-rules-javaee,2.7.0.Final" />
+            <addon id="org.jboss.windup.rules,windup-rules-java,2.7.0.Final" />
+        </dependencies>
+        <phase>PostMigrationRulesPhase</phase>
+    </metadata>
+    <rules>
+        <rule id="technology-usage-database-01000">
+            <when>
+                <graph-query discriminator="JdbcDatasourceModel"  />
+            </when>
+            <perform>
+                <technology-identified name="JDBC datasources">
+                    <tag name="Store"/>
+                    <tag name="Database"/>
+                    <tag name="Java EE"/>
+                </technology-identified>
+            </perform>
+        </rule>
+        <rule id="technology-usage-database-01001">
+            <when>
+                <graph-query discriminator="DataSourceModel">
+                    <property name="isXA">true</property>
+                </graph-query>
+            </when>
+            <perform>
+                <technology-identified name="JDBC XA datasources">
+                    <tag name="Store"/>
+                    <tag name="Database"/>
+                    <tag name="Java EE"/>
+                </technology-identified>
+            </perform>
+        </rule>
+        <rule id="technology-usage-database-01100">
+            <when>
+                <graph-query discriminator="JPAEntityModel" />
+            </when>
+            <perform>
+                <technology-identified name="JPA entities">
+                    <tag name="Store"/>
+                    <tag name="Persistence"/>
+                    <tag name="Java EE"/>
+                </technology-identified>
+            </perform>
+        </rule>
+        <rule id="technology-usage-database-01200">
+            <when>
+                <graph-query discriminator="JPANamedQueryModel"  />
+            </when>
+            <perform>
+                <technology-identified name="JPA named queries">
+                    <tag name="Store"/>
+                    <tag name="Persistence"/>
+                    <tag name="Java EE"/>
+                </technology-identified>
+            </perform>
+        </rule>
+        <rule id="technology-usage-database-01300">
+            <when>
+                <graph-query discriminator="JPAPersistenceUnitModel"  />
```

Comment:
This rules is failing the test because when it's executed, the method [TechnologyIdentified.perform](https://github.com/windup/windup/blob/master/rules-java-ee/addon/src/main/java/org/jboss/windup/rules/apps/javaee/TechnologyIdentified.java#L130) throws the exception `The method org.jboss.windup.graph.model.HasApplications.getApplications has no annotations, therefore frames cannot handle the method.` at [line 130](https://github.com/windup/windup/blob/master/rules-java-ee/addon/src/main/java/org/jboss/windup/rules/apps/javaee/TechnologyIdentified.java#L130).
I solved adding, at the end of [JPAPersistenceUnitModel](https://github.com/windup/windup/blob/master/rules-java-ee/addon/src/main/java/org/jboss/windup/rules/apps/javaee/model/JPAPersistenceUnitModel.java#L83) class, the following code:
```java
    @JavaHandler
    @Override
    Iterable<ProjectModel> getApplications();

    @JavaHandler
    @Override
    boolean belongsToProject(ProjectModel projectModel);

    abstract class Impl implements JPAPersistenceUnitModel, JavaHandlerContext<Vertex>
    {
        public Iterable<ProjectModel> getApplications()
        {
            return this.getJPAConfigurationFileModel().getApplications();
        }

        public boolean belongsToProject(ProjectModel projectModel)
        {
            return this.getJPAConfigurationFileModel().belongsToProject(projectModel);
        }
    }
```

Reply:
Nice catch. I'm not sure why I didn't get that exception :( Maybe I missed that...

Reply:
I saw the exception only when i debugged the rule to understand why it was not working as expected.